Reorganise the devguide to directories

AA-Turner opened the pull request:
From the discord discussion, this is an attempt to add some structure to Docs/Tests/Triage and clean up the top level.

No content has been edited, except for combing two duplicated "how to become a triager" sections. As I moved content between pages I had to edit the underlines -- I also standardised some `^` underlines and out-of-order underlines to allow easier page moves in the future (e.g. a core-dev folder, buildbots, language design and change).

cc: @ezio-melotti @encukou @CAM-Gerlach 

Preview: https://cpython-devguide--901.org.readthedocs.build/

Follows on from #895 where things went catastrophically wrong.

A

brettcannon commented:
FYI I have no opinion on this, so I'm dropping my review request.

CAM-Gerlach commented:
FYI, just in time for this PR, GitHub (finally) implemented the ability to [seamlessly view commit history across file renames and moves](https://github.blog/changelog/2022-06-06-view-commit-history-across-file-renames-and-moves/).

CAM-Gerlach commented on triaging/labels.rst:
```diff
@@ -310,63 +396,18 @@ a link to relevant web pages.
-.. _Python source code repositories: https://github.com/python/cpython/
 .. _Reporting security issues in Python: https://www.python.org/dev/security/
 .. _python-ideas: https://mail.python.org/mailman3/lists/python-ideas.python.org
 .. _release schedule: https://devguide.python.org/#status-of-python-branches
```
This is one place, but I didn't make a suggestion since the actual usage isn't in the scope I can comment on here

CAM-Gerlach replied:
```suggestion
```

This was fixed above and marked resolved but missed being removed here.

CAM-Gerlach commented on developer-workflow/communication-channels.rst:
```diff
@@ -1,14 +1,16 @@
 
 Standards of behaviour in these communication channels
-------------------------------------------------------
+======================================================
```
```suggestion
======================================================

```

CAM-Gerlach commented on developer-workflow/communication-channels.rst:
```diff
@@ -111,7 +113,7 @@ It is also a common venue for the core developer promotion votes (this
 
 Tutorials for new users
-'''''''''''''''''''''''
+-------------------------
```
```suggestion
-------------------------

```

CAM-Gerlach commented on developer-workflow/communication-channels.rst:
```diff
@@ -151,10 +153,10 @@ bottom progress bar to expand it.
 
 Following categories (Category notifications)
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+'''''''''''''''''''''''''''''''''''''''''''''
```
```suggestion
'''''''''''''''''''''''''''''''''''''''''''''

```

CAM-Gerlach commented on developer-workflow/communication-channels.rst:
```diff
@@ -166,21 +168,21 @@ All categories are set by default in Normal mode where you will only be notified
 
 Following individual threads (Topic notifications)
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+''''''''''''''''''''''''''''''''''''''''''''''''''
```
```suggestion
''''''''''''''''''''''''''''''''''''''''''''''''''

```

CAM-Gerlach commented on developer-workflow/communication-channels.rst:
```diff
@@ -166,21 +168,21 @@ All categories are set by default in Normal mode where you will only be notified
 
 Customising notifications on user preference
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+''''''''''''''''''''''''''''''''''''''''''''
```
```suggestion
''''''''''''''''''''''''''''''''''''''''''''

```

CAM-Gerlach commented on developer-workflow/communication-channels.rst:
```diff
@@ -166,21 +168,21 @@ All categories are set by default in Normal mode where you will only be notified
 
 Enabling mailing list mode
-''''''''''''''''''''''''''
+--------------------------
```
```suggestion
--------------------------

```

CAM-Gerlach reviewed (commented):
Thanks again, @AA-Turner ! LGTM, aside from a few minor issues.

* [x] I grepped and found hardcoded external links to devguide.python.org three places (which are now outdated), which should be replaced with `:ref:` and any links using them updated:
    * [x] ``core-developers/committing.rst:231:.. _`Python Triage Team` ``
    * [x] ``getting-started/getting-help.rst:29:  `Discourse <https://devguide.python.org/communication/#discourse-discuss-python-org-web-forum>`_``
    * [x] ``triaging/labels.rst:413:.. _release schedule: https://devguide.python.org/#status-of-python-branches``

* [x] Also, I'm still curious about the under/over header format used here—while an extra column on each side is included in some of the [docutils examples](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#sections), it is not what is specified in either the [Sphinx documentation](https://devguide.python.org/documenting/#sections) or the canonical [Python docs style guide](https://devguide.python.org/documenting/#sections) in this very repo, nor have I commonly seen it elsewhere...

hugovk commented on advanced-tools/clang.rst:
```diff
@@ -1,6 +1,8 @@
-***************************
+.. _clang:
+
+=============================
+ Dynamic Analysis with Clang
+=============================
```
This form with added spaces looks a little unusual. Should we keep it flush without spaces? This is how the devguide recommends it:

```suggestion
===========================
Dynamic Analysis with Clang
===========================
```

(Same for [PEPs](https://peps.python.org/pep-0012/#section-headings).)

(And without spaces would cause less churn when promoting/demoting section.)

CAM-Gerlach replied:
Yeah, I likewise don't understand the mass change, especially given it is not what is currently used anywhere in the devguide itself either and thus causes a lot of noisy changes to files that would otherwise be cleanly moved, or nearly so, and is also more needlessly complicated and nitpicky to get right.

As discussed above:

> Also, I'm still curious about the under/over header format used here—while an extra column on each side is included in some of the [docutils examples](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#sections), it is not what is specified in either the [Sphinx documentation](https://devguide.python.org/documenting/#sections) or the canonical [Python docs style guide](https://devguide.python.org/documenting/#sections) in this very repo, nor have I commonly seen it elsewhere...

I also commented and suggested such on the previous review, but they were all marked as resolved without any explanation.

ezio-melotti replied:
FWIW I also agree that there shouldn't be any extra leading/trailing column.

ezio-melotti commented on advanced-tools/index.rst:
```diff
@@ -0,0 +1,10 @@
+================
+
+.. toctree::
+   :maxdepth: 1
```
Since the index pages are mostly empty, wouldn't it be better to add more depth?
This will provide a better overview of the pages contents (since it might not be obvious which page contains what you are looking for), and provide direct links to specific sections.

ezio-melotti commented on appendix.rst:
```diff
@@ -1,70 +0,0 @@
-Appendix: Topics
```
Is this file completely gone?
The file itself is probably redundant after this PR, but are the "Recommended readings" preserved somewhere?

AA-Turner replied:
> are the "Recommended readings" preserved somewhere?

No. Where would you suggest, if we should keep them?

A

CAM-Gerlach replied:
As @ezio-melotti suggests right below, they could go under the Audience categories (Contributor, Triager, Documentarian, CoreDev) in the restored table in the index. That would be a lot better than reverting whole removal and is the same or less work than reverting the removal since it would mean updating the doc links to ref links and fixing the target names, or fixing the outdated file paths and names, and also adding the pages that were added/split and now missing here or in the wrong sections.

AA-Turner replied:
> updating the doc links to ref links and fixing the target names, or fixing the outdated file paths and names, and also adding the pages that were added/split and now missing here or in the wrong sections.

I did these in a commit that I kept, so they should still work. Time-wise I probably don't have time to go through and update the table in the next week or so, so reverting was the quickest option.

A

CAM-Gerlach replied:
For the table (as far as I can tell), but not the appendix, at least as of the version you've pushed here—it still uses `:doc:` links, and almost all of them are now broken, as can be seen [on the preview](https://cpython-devguide--901.org.readthedocs.build/appendix/). Furthermore, it is missing the documents split as part of this PR, leaving it incomplete.

Instead of tediously updating all of that (or introducing a major regression), only to remove it again in a followup PR, I suggest just spending the same or less time re-dleeting it and adding them to the table instead as nessesary. Additionally, to keep it shorter and less work to update, we can just link the top-level index pages of sections that are obviously of central relevance to the audience type in their entirety, instead of linking every sub-page.

 I'm happy to suggest an outline for it, or make a PR to your branch to implement it if you don't have time. Here's what I recommend:

<details>

<summary>Suggested table structure (by column heading)</summary>

All contributors:
* Getting Started
* Following Python's Development
* Issue Tracker
* Running and Writing Tests

Documentarians:
* Documentation
* Development Cycle
* Experts Index
* Triaging an Issue

Triagers:
* Issues and Triaging
* Development Workflow
* Experts Index
* Accepting Pull Requests

Core Developers:
* Core Developers
* Development Workflow
* Issues and Triaging
* Testing and Buildbots

</details>

ezio-melotti commented on core-developers/developers.csv:
```diff
@@ -1,180 +1,180 @@
-Erlend Egeberg Aasland,erlend-aasland,2022-05-05,,
```
Did the line endings change or is it just git getting confused by the move?
This file is generated, so it shouldn't be changed.

In addition, I'm planning to move this in an `include` directory as part of #884.

AA-Turner replied:
> just git getting confused by the move?

This, the line endings shouldn't've changed.

A

AA-Turner replied:
Reverted.

A

ezio-melotti replied:
Thanks for reverting!  As I mentioned that file is generated, so reverting the line endings change is the right thing to do, or the next update to the file will result in a big diff that will revert them all.

ezio-melotti commented on core-developers/developer-log.rst:
```diff
@@ -1,3 +1,4 @@
+.. _developer-log:
 .. _developers:
```
Why two targets?

AA-Turner replied:
Keeps the old target working (for `#fragments`) and ensures the filename is always a target as a simple rule-of-thumb.

A

ezio-melotti commented on documenting/style-guide.rst:
```diff
@@ -0,0 +1,223 @@
+This document describes the style guide for our documentation.
+
+
+Use of whitespace
```
```suggestion
Use of Whitespace
```

hugovk replied:
The style guide currently uses a mix of "Sentence case" and "Title Case" for section headers.

Should we rather update the style guide to follow _itself_ and primarily use sentence case? :)

https://devguide.python.org/documenting/#capitalization

ezio-melotti commented on getting-started/pull-request-lifecycle.rst:
```diff
@@ -1,12 +1,15 @@
+.. _pullrequest:
 
+=============================
 Lifecycle of a Pull Request
```
This is off-center according to this PR's header style, but if we revert back to the previous style it doen't matter.

ezio-melotti commented on index.rst:
```diff
@@ -154,38 +154,7 @@ Core developers and contributors alike will find the following guides useful:
-| :doc:`communication`   |                     |                       | :ref:`office hour`  |
-+------------------------+---------------------+-----------------------+---------------------+
-| :doc:`gitbootcamp`     |                     |                       |                     |
-+------------------------+---------------------+-----------------------+---------------------+
```
This table grouped the page by intended audience, which is somewhat different from the grouping implemented in the sidebar.  I think it might be better to keep it and possibly update it.
Here it would also be a good place to list the "recommended readings" that were suggested in the appendix.

CAM-Gerlach replied:
That could make sense, and is essentially the approach I was planning for a followup PR if the rest of the restructuring wasn't done here. One idea: Instead of listing every individual page, which is way longer and a lot more work to maintain and risk of desync, we could just list the category index pages if most or all of the pages in the category apply.

ezio-melotti commented on index.rst:
```diff
@@ -154,38 +154,7 @@ Core developers and contributors alike will find the following guides useful:
-| :doc:`gitbootcamp`     |                     |                       |                     |
-+------------------------+---------------------+-----------------------+---------------------+
-
-Advanced tasks and topics for once you are comfortable:
-
-* :doc:`silencewarnings`
-* Fixing issues found by the :doc:`buildbots <buildbots>`
-* :doc:`coverity`
-* Helping out with reviewing `open pull requests`_.
-  See :ref:`how to review a Pull Request <how-to-review-a-pull-request>`.
-* :doc:`fixingissues`
```
This part is no longer necessary.

ezio-melotti commented on internals/garbage-collector.rst:
```diff
@@ -1,14 +1,17 @@
+.. _garbage-collector:
 .. _gc:
+.. _garbage_collector:
```
Are these for backward-compatibility with existing links?

CAM-Gerlach replied:
Yes

ezio-melotti commented on internals/garbage-collector.rst:
```diff
@@ -1,14 +1,17 @@
+ Garbage Collector Design
+==========================
 
 :Author: Pablo Galindo Salgado
```
Shouldn't this be lowercase or is it case-insensitive?

AA-Turner replied:
> this be lowercase or is it case-insensitive?

Which bit are you referring to? The title or the author field?

A

AA-Turner replied:
I didn't change it in this PR, so I'll mark as resolved.

A

ezio-melotti commented on triaging/triage-team.rst:
```diff
@@ -0,0 +1,100 @@
+.. _PSF Code of Conduct: https://www.python.org/psf/conduct/
+
+
+Becoming a member of the Python triage team
```
Should this use title-case?
```suggestion
Becoming a Member of the Python Triage Team
```

AA-Turner commented on internals/garbage-collector.rst:
```diff
@@ -181,7 +184,7 @@ of that object when the algorithm starts. This is because the algorithm needs
 interpreter will not modify the real reference count field.
 
-.. figure:: images/python-cyclic-gc-1-new-page.png
+.. figure:: /_static/python-cyclic-gc-1-new-page.png
```
I'd say different PR. (I am *trying* to keep scope limited!)

A

AA-Turner commented:
I do have a concern on scope -- for instance title case / sentence case of headings. I would suggest to merge-as-is (once I've resolved the outstanding points) and do those sort of things as follow-ups.

A

AA-Turner commented:
Fixed references, increased max depth, removed spaces from titles, reverted line ending change in developers.csv.

A

AA-Turner commented:
I restored the appendix and topic table again to reduce scope of this PR as it seems removing them needs more discussion (Ezio's points). 

Partially I want to get this in as it puts this / other PRs in a state of limbo w.r.t. the moves.

A

CAM-Gerlach reviewed (commented):
Minor note: The three remaining instances of the devguide links using external URLs were updated in the text, but the link targets were not actually removed. Searching for `.. _\w+.*: .*devguide\.python` should find all of them.

While I agree we shouldn't scope-creep this further and change the style of all the headings, I don't think we should rush this PR now without resolving the final issue with the appendix and index table being redundant/out of date, since:

* These are pretty central to the question of how to organize and structure the devguide (more so than some of the other implemented changes)
* This would result in a significant regression since the appendix is missing the newly split docs, has out of date file paths and is either out of sync with or duplicate much of the new structure, unless we do the work here to update it.
* There seems to be a fairly straightforward solution suggested by multiple people here, which would probably be around the same amount of work as updating the existing redundant appendix.

Also, keep in mind this can't be merged immediately until someone with the appropriate RTD permissions sets up the redirects (or we do them ourselves with my script or `sphinx-redirects`), _and_ a repo admin can merge the PR (due to the CLA bot bug)— @ezio-melotti are you one here? And the existing PRs _should_ transfer, so long as the move was detected (as most were) instead of the page split, and the relevant text wasn't changed.

As @ezio-melotti suggests, the various recommended reading relevant to each of the audience types can go under the respective table colums (Contributor, Triager, Documentarian, CoreDev) in the table on the index page index. That would be a lot better than reverting whole appendix removal and should be a similar amount of work as updating both to properly reflect the changes.

hugovk commented:
> Also, keep in mind this can't be merged immediately until someone with the appropriate RTD permissions sets up the redirects (or we do them ourselves with my script or `sphinx-redirects`), _and_ a repo admin can merge the PR (due to the CLA bot bug)— @ezio-melotti are you one here?

Some smallish benefits to having Sphinx-level redirects (either your script, `sphinx-redirects`, `sphinx-reredirects`, `sphinxext-rediraffe`, ...) _in addition_ to more-efficient RTD ones:

* Redirects go live immediately, don't need to sync with setting RTD config
* We get redirects on our forks on RTD (without having to also set up the RTD config)
* We get redirects for local testing
* They're defined in code (although it would a good to put the RTD config under source control for transparency, copying and pasting)

CAM-Gerlach commented:
I definitely agree; having them in the repo and generating them as part of the build process allows testing them locally, is portable between hosts and deployment mechanisms, and makes 100% sure they are in sync with the repo itself (particularly in this case, since otherwise a bunch of inbound links will break), among the other benefits you mentioned.

[`sphinxext-rediraffe`](https://sphinxext-rediraffe.readthedocs.io/en/latest/) seems really powerful and well suited to this use case, as it can resolve chained redirects while only actually doing a single redirect client-side, can check that deleted/moved pages in the GitHub history are directed, and is an integrated Sphinx extension. Sphinx-reredirect seems to be the other option; its simpler and a bit fresher in terms of maintenance, but doesn't seem to be nearly as sophisticated as rediraffe, especially with its handy features well suited for this use case.

Of course, the RTD config should also be checked in to the repo, especially on one with many contributions/contributors and few admins with limited bandwidth; I've found in general that always checking in infra config and updating them in sync with PRs is far more maintainable than buried in the config of some external service asynchronous from the content changes and that only a few people have access to.

hugovk commented:
Good news, the CLA bot is fixed! We can do a regular merge when we're ready!

pradyunsg commented:
Here's an example of how to have HTML meta redirects without adding a new extension or such, within regular reStructuredText.

https://raw.githubusercontent.com/pypa/pip/main/docs/html/reference/pip_cache.rst

CAM-Gerlach commented:
> Here's an example of how to have HTML meta redirects without adding a new extension or such, within regular reStructuredText.

Sure, it could be done manually, but its simpler (one line of config, or even automatic, vs. several lines of boilerplate), less clutter (avoids lots of stub files in the source) and more efficient (the generated pages can be much shorter and only contain `<head>` so they don't have to wait to load and render) to do it with an extension, as well as allows for checking automatically and on-demand for any page moves not covered by redirects and auto-generating redirects for them, and I'm not sure anyone's raised an objection to adding one.

I'm also not sure why in the example, the page makes the user sit and wait for three seconds to the content they want instead of simply redirecting instantly, making the redirect as transparent as possible to the user, which seems to increase the disruption.

> Given that all of the newly introduced "root" documents are basically just toctrees, I think an option worth considering is:

>    Not moving the pages around, and only changing the organisation in the Sphinx toctree directives.

> I haven't looked at the actual code changes, only the rendered documentation, but IMO this approach would also lend toward an easier-to-review PR -- since none of the files would move, none of the pages need updates to their references and the markup changes/cleanups can move into a dedicated follow up. 

This would have been a good idea as a first step if we'd done this to begin with, but for better or worse all those changes have already been made, and it would require substantially more work to go back and rewrite everything now. And a number of pages would still have to be moved, as these changes don't just move around existing pages, but also split, combine and re-organize a number of them into better organized and more manageable documents—so we'd need to have the redirects anyway, and it is those moves that are more difficult to review and can cause merge conflicts, and that's the case regardless of whether we have a flat or nested structure; the others are tracked by GitHub just fine.

Furthermore, since a lot of their names gotten out of sync with their actual contents, or are now, or are confusingly similar to others (making it hard to navigate the source and by URLs), they've also been renamed while being moved anyway. On the whole, dumping everything in one flat directory works well for smaller projects (its what I used to do in Spyder's docs, where we used unlinked toctrees to give things a structure), but can become really unmanageable and hard to navigate once things get near this size, and both the cost of keeping it and the cost of switching will only increase over time, whereas adding Sphinx-Rediraffe and switching to `:ref:` rather than `:doc:` to link them is a one-time cost (paid in this PR) and transparent to users, and will make any future moves transparent to devguide contributors as well.

> Having multiple toctrees with a caption on each, for grouping the various documents -- Furo's own documentation has a toctree with a Development caption: [pradyunsg/furo@main/docs/index.md (raw)](https://raw.githubusercontent.com/pradyunsg/furo/main/docs/index.md)

It would definitely be nice to have toctree section headers that open when clicked rather than linked to the toctree page, [like we have in the Spyder docs in our custom theme](https://docs.spyder-ide.org/current/index.html). But in that case (maybe there's a Furo option for it?) those headings aren't collapsed, meaning all 30-40 documents would be listed in the sidebar at once (plus the category names), which is overwhelming and hard to navigate, and one of the motivating reasons for the change—something I experienced first hand as a newer contributor, and even now as an experienced one with the current devguide vs. using the rendered preview of this PR.

AA-Turner commented:
Having the logical organisation differ from the filesystem makes things harder to find when wanting to contribute -- the real ideal here would be for Furo to add a config line to have blank index pages -- I could override the theme and add it, but I'd really rather not.

I'd posit the second best is this PR, as the index pages are de-emphasised as far as possible.

A

AA-Turner commented:
I've updated the topic table and re-removed the appendix, hopefully this resolves your concerns @ezio-melotti @CAM-Gerlach?

Are there any other outstanding points on this PR?

A

hugovk commented:
Outstanding: redirects from deleted pages to new ones; ideally local redirects, and also a text file or something with the RTD config for easy copy and pasting by an admin.

ezio-melotti commented on index.rst:
```diff
@@ -156,36 +156,27 @@ Core developers and contributors alike will find the following guides useful:
-It is **recommended** that the above documents be read as needed. New
-contributors will build understanding of the CPython workflow by reading the
-sections mentioned in this table. You
+======================== =================== ======================= =======================
+Contributors             Documentarians      Triagers                Core Developers
+======================== =================== ======================= =======================
+:ref:`setup`             :ref:`docquality`   :ref:`tracker`          :ref:`responsibilities`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`help`              :ref:`documenting`  :ref:`triaging`         :ref:`developers`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`pullrequest`       :ref:`style-guide`  :ref:`helptriage`       :ref:`committing`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`runtests`          :ref:`rst-primer`   :ref:`experts`          :ref:`devcycle`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`fixingissues`      :ref:`translating`  :ref:`labels`           :ref:`motivations`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`communication`                         :ref:`gh-faq`           :ref:`office hour`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`gitbootcamp`                           :ref:`triage-team`      :ref:`experts`
+------------------------ ------------------- ----------------------- -----------------------
+:ref:`devcycle`
+------------------------ ------------------- ----------------------- -----------------------
```
```suggestion
======================== =================== ======================= =======================
Contributors             Documentarians      Triagers                Core Developers
======================== =================== ======================= =======================
:ref:`setup`             :ref:`docquality`   :ref:`tracker`          :ref:`responsibilities`
:ref:`help`              :ref:`documenting`  :ref:`triaging`         :ref:`developers`
:ref:`pullrequest`       :ref:`style-guide`  :ref:`helptriage`       :ref:`committing`
:ref:`runtests`          :ref:`rst-primer`   :ref:`experts`          :ref:`devcycle`
:ref:`fixingissues`      :ref:`translating`  :ref:`labels`           :ref:`motivations`
:ref:`communication`                         :ref:`gh-faq`           :ref:`office hour`
:ref:`gitbootcamp`                           :ref:`triage-team`      :ref:`experts`
:ref:`devcycle`
======================== =================== ======================= =======================
```

ezio-melotti commented:
There are a few warnings in https://github.com/python/devguide/runs/7053952512?check_suite_focus=true
My suggestion above should fix (and simplify) the table.

AA-Turner commented:
> and also a text file or something with the RTD config for easy copy and pasting by an admin.

I looked on the RTD interface for redirects and it seems there's no way to upload a file, you need to select the redirect type and the to / from for every single page. I could add such a list but given it can't be used directly in the RTD interface and we already have the mapping from `sphinx-rediraffe` I'm not sure what it would add.

A

AA-Turner commented:
Hmm, CI is failing as the devguide is built with dirhtml -- the redirects for `triaging` and `documenting` to their replacement pages don't work, as the pages sphinx-rediraffe wants to create to use for the redirects already exist. I've commented out those two for now to see if things are fixed.

A

CAM-Gerlach commented:
We could rename `triaging` to `issues` (as I originally suggested, and is arguably a more appropriate name), and likewise, we could rename `documenting` to `documentation`.

CAM-Gerlach reviewed (approved):
I tested the redirects and did a quick final smoke check, and everything LGTM now :+1: , aside from the issue discussed above with the `documenting` and `triaging` redirects, that could be solved by renaming the directories.

hugovk commented:
> > and also a text file or something with the RTD config for easy copy and pasting by an admin.
> 
> I looked on the RTD interface for redirects and it seems there's no way to upload a file, you need to select the redirect type and the to / from for every single page. I could add such a list but given it can't be used directly in the RTD interface and we already have the mapping from `sphinx-rediraffe` I'm not sure what it would add.
> 
> A

Indeed, as suggested in https://github.com/python/devguide/pull/901#issuecomment-1160048582, the text file is for:

* transparency: only the RTD admin can set the config in the interface, but a text file means we can also see what's there
* simplicity: makes it easier for the RTD admin to know what to set up in the RTD interface

The reason for HTTP redirects in RTD as well is they are more efficient than Sphinx-level ones. I think @CAM-Gerlach wrote something above or in https://github.com/python/devguide/pull/895 about it but can't find it now...

We could even skip RTD-level directs or add them later, don't necessarily need to hold this PR any further. As long as we at least have _some_ redirects ready (RTD or Sphinx) when this is merged to prevent 404s.

CAM-Gerlach commented:
> The reason for HTTP redirects in RTD as well is they are more efficient than Sphinx-level ones. I think @CAM-Gerlach wrote something above or in https://github.com/python/devguide/pull/895 about it but can't find it now...

They are certainly more efficient (I seem to recall you, Ee or one of us testing them on the PEPs repo perhaps?), though if the HTML redirect pages are as well optimized as we have them on the Spyder docs site with my custom script (minimal head/content to minimize download time, no assets to avoid blocking on extra requests, and no body to bypass any rendering), its a matter of perhaps a few hundred ms at most, which is never going to be as fast as a server or CDN-level redirect, but not a big deal in practice at least over the short term.

Maybe we could put the devguide behind Fastly as well, so it could be configured in the repo like we have it on the PEPs?

ezio-melotti commented:
FTR PEP redirects were discussed in https://github.com/python/peps/issues/2420

AA-Turner commented:
Updated to documentation/ and triage/ -- I don't want to use issues/ as it covers PRs as well.

A

ezio-melotti commented:
Merge commits are not enabled for this repo -- only squash and rebase merges.  If squashing makes the splitting/moving difficult to track for git, we could either rebase these 21 commits or the reformatted commits if you want to do that.  If it's too much work I'm fine with rebasing the 21 commits, unless someone else disagrees.

AA-Turner commented:
Rebase merge may be the best option then it seems, thanks for looking

A

hugovk commented:
Here's a `git merge --squash` on my fork:
* https://github.com/hugovk/devguide/commits/main2
* https://github.com/hugovk/devguide/commit/f89203eeee99013c2e1d7c1a554cc02000592f3f

CAM-Gerlach commented:
The main argument as I recall for doing it all in one monolithic PR was so it could be squashed down to one atomic commit, that could easily be skipped, ignored, etc, with minimum noise, which splitting it up would at least partially defeat.

The Files pane of this PR should already show what files are tracked as moved, versus seen as created/deleted, no? And indeed, as far as I can tell, it is identical to viewing the same in the merge commit that @hugovk helpfully linked.

It seems like pretty much the only files not detected as moved were those that were actually split or combined from other files, which wouldn't really be tracked any better at the end anyway—at most a single one file would be tracked all the way through. So it doesn't seem a non-squash merge would actually meaningfully benefit this much if at all.

> I don't want to use issues/ as it covers PRs as well.

Well, by that reasoning it covers a lot more than triaging too; the whole [issue tracker page](https://cpython-devguide--901.org.readthedocs.build/triage/issue-tracker/) doesn't cover triaging at all (its written for those submitting issues, not triaging them), and likewise the [GitHub Issues for BPO Users FAQ](https://cpython-devguide--901.org.readthedocs.build/triage/github-bpo-faq/) almost exclusively covers questions and tasks from the user perspective, that relate to navigating and creating issues rather than triaging.

But eh, it's really not _that_ important, heh :shrug:

ezio-melotti commented:
FWIW I'm planning to revise/rewrite the whole section about issues at some point.

hugovk commented on conf.py:
```diff
@@ -226,6 +231,44 @@
+    "clang.rst": "advanced-tools/clang.rst",
+    "coverity.rst": "advanced-tools/coverity.rst",
+    "gdb.rst": "advanced-tools/gdb.rst",
+    "coredev.rst": "core-developers/becoming-coredev.rst",
```
And rename the file:

```suggestion
    "coredev.rst": "core-developers/becoming-core-dev.rst",
```

hugovk commented on core-developers/becoming-coredev.rst:
```diff
@@ -0,0 +1,59 @@
+.. _becoming-coredev:
```
```suggestion
.. _becoming-core-dev:
```

hugovk commented on core-developers/index.rst:
```diff
@@ -0,0 +1,13 @@
+   experts
+   developer-log
+   motivations
+   becoming-coredev
```
```suggestion
   becoming-core-dev
```

hugovk reviewed (approved):
Looks good, just one little naming nit.

ezio-melotti reviewed (commented):
In `index.rst` at line 113 there's a reference to `devcycle.rst` in a comment that should be updated too.

ezio-melotti commented:
Perhaps out of the scope of this PR, but in https://github.com/endoflife-date/endoflife.date/discussions/711 it was suggested to move the "Status of the Python branches" and the "End-of-life branches" tables on a page on their own.  In #884 I'm refactoring them to use `csv`s`, and if we agree on moving those two on their own page we should do it shortly after this PR is merged.

CAM-Gerlach commented:
> Perhaps out of the scope of this PR, but in https://github.com/endoflife-date/endoflife.date/discussions/711 it was suggested to move the "Status of the Python branches" and the "End-of-life branches" tables on a page on their own.

Up to you two whether its in scope here, but that would be a great idea, IMO—just recently I was checking those pages and noticed how much scrolling it took to find the actual table.

ezio-melotti commented:
> Up to you two whether its in scope here

Technically it could, but I don't want to block this further.  We can create a new page in a separate follow-up PR once this is merged.

AA-Turner commented:
Renamed `becoming-coredev` to `become-core-developer` and added a new top-level `versions` page.

Thanks to Hugo for trying the squash merge -- it isn't as bad as I feared but there are some page moves that get incorrectly attributed (e.g. `coredev` -> `responsibilities` and `documenting` -> `markup`) so I'd prefer the rebase-merge if possible.

A

ezio-melotti commented:
We can decide in the meeting whether to squash or rebase, but otherwise it looks good.

AA-Turner commented:
Will update now ahead of the meeting.

A

AA-Turner commented:
Done.

A
